lxd/network/network_utils.go

```diff
@@ -639,6 +639,109 @@ func randomSubnetV6() (string, error) {
 	return "", fmt.Errorf("Failed to automatically find an unused IPv6 subnet, manual configuration required")
 }
 
+// randomAddressInSubnet finds a random address within a given subnet. If given, the validate function is called for
+// each randomly generated value. Each time validate returns an error, another IP address will be generated until the
+// context is cancelled.
+func randomAddressInSubnet(ctx context.Context, ipnet *net.IPNet, validate func(net.IP) error) (net.IP, error) {
+	var bufferLen int
+	var ipVersion int
+	if ipnet.IP.To4() != nil {
+		ipVersion = 4
+		bufferLen = 4
+	} else {
+		ipVersion = 6
+		bufferLen = 16
+	}
+
+	ones, _ := ipnet.Mask.Size()
+
+	// The number of bytes that will be the same in the subnet and in the random address.
+	quotient := ones / 8
+
+	// The remaining bits that will be the same in the subnet and in the random address (after the equivalent bytes).
+	remainder := ones % 8
+
+	// Can't get a random address from e.g. 192.0.2.1/32 as it's a single IP.
+	if quotient == bufferLen {
+		return nil, fmt.Errorf("Cannot find random address when subnet contains only one address")
+	}
+
+	for {
+		// Exit if the context is cancelled.
+		select {
+		case <-ctx.Done():
+			return nil, ctx.Err()
+		default:
+		}
+
+		// Read randomness into a byte slice of the correct length for the IP version.
+		r := make([]byte, bufferLen)
+		_, err := cryptorand.Read(r)
```

Comment:
Could we simplify this process by taking the starting number of the subnet and then using https://pkg.go.dev/crypto/rand#Int to get a random number between 0 and the max size of the subnet and then adding it to the subnet?

Reply:
This is a bit like what getDHCPFreeIPv4() does in the dnsmasq dhpcalloc package, except its adding 1 each time.